torch/csrc/jit/interpreter.cpp

```diff
@@ -1,42 +1,208 @@
+#include "Python.h"
 #include "interpreter.h"
 #include "torch/csrc/jit/ir.h"
 #include "torch/csrc/autograd/profiler.h"
 #include "torch/csrc/jit/generated/aten_dispatch.h"
+#include "Python.h"
+#include "pybind11/pybind11.h"
+#include "torch/csrc/utils/auto_gil.h"
+#include "torch/csrc/autograd/variable.h"
+#include "torch/csrc/autograd/python_variable.h"
+#include "torch/csrc/autograd/python_engine.h"
+#include "torch/csrc/autograd/functions/special.h"
 #ifdef WITH_CUDA
 #include "torch/csrc/jit/fusion_compiler.h"
 #endif
 
+namespace py = pybind11;
+
 namespace torch { namespace jit {
 
+struct DummyFunction : autograd::Function {
+  DummyFunction() {
+    num_inputs = 0;
+    is_executable = true;
+  }
+  virtual autograd::variable_list apply(const autograd::variable_list& inputs) override {
+    throw std::logic_error("DummyFunction::apply() called, but it should be blocked by a callback returning false");
+  }
+};
+
+struct Handle : at::RefCounted {
+  std::shared_ptr<DummyFunction> forward_inputs;
+  autograd::function_list forward_outputs;
+};
+
+at::Tensor unsafeTensorBorrow(at::RefCounted * rc) {
+  return at::Tensor(static_cast<at::TensorImpl*>(rc), true);
+}
+
+struct HandleBuilder {
+  HandleBuilder(bool requires_handle) {
+    if(requires_handle) {
+      handle = new Handle();
+      handle->forward_inputs = std::make_shared<DummyFunction>();
+    }
+  }
+  autograd::Variable addInput(at::RefCounted* input, const VariableFlags & flags_) {
+    if(handle && flags_.requires_grad) {
+      autograd::VarFlags flags = {flags_.requires_grad, flags_.is_volatile};
+      return autograd::make_variable(
+        unsafeTensorBorrow(input),
+        flags,
+        handle->forward_inputs->num_inputs++,
+        handle->forward_inputs);
+    } else {
+      autograd::VarFlags flags = {false, false};
+      return autograd::make_variable(unsafeTensorBorrow(input), flags);
+    }
+  }
+  at::RefCounted* addOutput(const autograd::Variable & output) {
+    if(handle) {
+      handle->forward_outputs.emplace_back(output.grad_fn(),output.output_nr());
+    }
+    at::Tensor tensor = output.data();
+    return tensor.detach();
+  }
+  void writeTo(refcounted_list & outputs) {
+    // note: no if(handle) guard
+    // because an unused handle is still produced as an output
+    outputs.push_back(handle);
+  }
+private:
+  Handle* handle = nullptr;
+};
+
+bool hasHandleOutput(Node * n) {
+  if(n->outputs().size() == 0)
+    return false;
+  auto & last = n->outputs().back();
+  return last->isHandle() && last->uses().size() > 0; // don't bother creating a handle if it is never used
+}
+
+Operation createPythonCallback(PythonOp* op) {
+  py::object func = py::handle(op->pyobj.get()).attr("apply");
+  bool has_handle = hasHandleOutput(op);
+  return [=](const refcounted_list & inputs, refcounted_list & outputs) {
```

Comment:
This implementation adds an important constraint that wasn't true before - we have to retain the JIT graph even after it was compiled, because you need the `op` pointer at run time! We should document this somewhere (probably in `python_compiled_function.cpp`, next to the `graph_` field).

Reply:
Added docs in jit::Code to remind people not to remove the graph handle.